x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,290 @@
+  private readonly httpAgent: HTTPAgent;
+  private readonly allowUnauthAgent: HTTPAgent;
+  private readonly wreck: typeof Wreck;
+  private readonly log: Logger;
+  private readonly config$: Observable<ProxyConfig>;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.nodeName = ''; // todo make this work!
+    this.config$ = initializerContext.config.create<ProxyConfig>();
+    this.log = initializerContext.logger.get('proxy');
+    this.httpsAgent = new HTTPSAgent({ keepAlive: true });
+    this.httpAgent = new HTTPAgent({ keepAlive: true });
+    this.allowUnauthAgent = new HTTPAgent({ keepAlive: true });
+    this.wreck = Wreck.defaults({
+      agent: {
+        https: this.httpsAgent,
+        http: this.httpAgent,
+        httpsAllowUnauthorized: this.allowUnauthAgent,
+      },
+    });
+  }
+
+  public async setup(core: CoreSetup) {
+    // initialize cluster information
+    // we need to get this kibana instance's ip address
+    // it's hostname and determine which port to listen on
+    // should liveness timing be part of configuration?o
+    // mikhael -- talk about changing http service to allow multiple services
+    this.configSubscription = this.config$.subscribe(config => {
+      this.setConfig(config);
+    });
+
+    const config = await this.config$.pipe(first()).toPromise();
+    this.setConfig(config);
+    this.elasticsearch = core.elasticsearch.dataClient$;
+    this.proxyServer = await core.http.createNewServer({ port: config.port });
+  }
+
+  private setConfig(config: ProxyConfig) {
+    this.updateInterval = config.updateInterval || this.updateInterval;
+    this.timeoutThreshold = config.timeoutThreshold || this.timeoutThreshold;
+    this.port = config.port || this.port;
+  }
+
+  public async start(core: CoreStart) {
+    // start up an http service
+    // update es doc to explain that we exist
+    // start updateLivenessMap
+    this.mainLoop();
+  }
+
+  public async stop() {
+    // stop http service
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+
+    const nodes = await this.getNodeList();
+    nodes.delete(this.nodeName);
+    await this.updateNodeList(nodes);
+
+    if (this.configSubscription === undefined) {
+      return;
+    }
+
+    this.configSubscription.unsubscribe();
+    this.configSubscription = undefined;
+  }
+
+  public async assignResource(resource: string, data: RoutingNode): Promise<void> {
+    this.routingTable.set(resource, data);
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async unassignResource(resource: string) {
+    this.routingTable.delete(resource);
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async registerRouter(router: Router) {
```

Comment:
@joshdover @restrry would love some opinions here on this section.

since the proxy needs to run it's own http server, the plugins will need to be able to register routes against the proxy's http server. is this an appropriate way to do so? it seems to match the public interface for how the main server does it?

@spacedragon @zfy0701 would love your opinions here as well

Reply:
The API itself looks fine to me, but no plugins are going to be able to access these methods written this way. They can only access whatever values are returned by your `setup` and `start` functions. Whatever is returned from these methods constitute your plugin's public API. Everything else is internal only at runtime.

Reply:
It could be helpful to have this class implement the `Plugin` interface which helps with typing the parameters and return values of the lifecycle functions.